tidy3d/components/data/sim_data.py

```diff
@@ -145,7 +142,42 @@ def source_spectrum_fn(freqs):
             # remove user defined phase from normalization so its effect is present in the result
             return spectrum * np.conj(user_defined_phase)
 
-        return monitor_data.normalize(source_spectrum_fn)
+        return source_spectrum_fn
+
+    def renormalize(self, normalize_index: int) -> SimulationData:
+        """Return a copy of the :class:`.SimulationData` with a different source used for the
+        normalization."""
+
+        if normalize_index == self.normalize_index:
+            # already normalized to that index
+            return self.copy()
+
+        num_sources = len(self.simulation.sources)
+
+        if num_sources == 0:
+            # no sources present
+            normalize_index = None
+            log.warning("No sources present in simulation, not doing normalization.")
+
+        if normalize_index and (normalize_index < 0 or normalize_index >= num_sources):
+            # normalize index out of bounds for source list
+            raise DataError(
+                f"normalize_index {normalize_index} out of bounds for list of sources "
+                f"of length {num_sources}"
+            )
+
+        def source_spectrum_fn(freqs):
+            """Normalization function that also removes previous normalization if needed."""
+            new_spectrum_fn = self.source_spectrum(normalize_index)
+            old_spectrum_fn = self.source_spectrum(self.normalize_index)
+            return new_spectrum_fn(freqs) / old_spectrum_fn(freqs)
+
+        # Make a new monitor_data dictionary with renormalized data
+        monitor_data = {}
+        for key, val in self.monitor_data.items():
+            monitor_data[key] = val.normalize(source_spectrum_fn)
+
+        return self.copy(update=dict(normalize_index=normalize_index, monitor_data=monitor_data))
```

Comment:
See my other comment in the original hdf5 performance PR, but doesn't this just defeat the purpose as in this line there are two `SimulationData` instances in memory at the same time? Maybe I'm just confused on this point: does a `return self.copy()` call hold two copies in memory? or is smart enough to free the memory for the original one before allocating for the returned one?

Reply:
So, I see your point, I would expect this to double the memory. However, the point of this PR is that in the vast majority of cases, people don't actually change the normalization: they just use the default `normalize_index = 0`. This is how the data will be provided from the server, so when they do `SimulationData.from_file`, no memory duplication will happen. I think the point is that while the less likely scenario of `sim_data.renormalize` does have memory doubling, this PR avoids memory doubling in the more likely scenario in which you have one very large monitor, and you do `sim_data["large_monitor"]`.